crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2059,6 +2114,293 @@ impl<'db> TypeInferenceBuilder<'db> {
+            }
+            ast::Expr::Set(set) => {
+                self.infer_set_expression(set);
+                Type::Unknown
+            }
+            ast::Expr::ListComp(listcomp) => {
+                self.infer_list_comprehension_expression(listcomp);
+                Type::Unknown
+            }
+            ast::Expr::SetComp(setcomp) => {
+                self.infer_set_comprehension_expression(setcomp);
+                Type::Unknown
+            }
+            ast::Expr::DictComp(dictcomp) => {
+                self.infer_dict_comprehension_expression(dictcomp);
+                Type::Unknown
+            }
+            ast::Expr::Generator(generator) => {
+                self.infer_generator_expression(generator);
+                Type::Unknown
+            }
+            ast::Expr::Await(await_expression) => {
+                self.infer_await_expression(await_expression);
+                Type::Unknown
+            }
+            ast::Expr::Yield(yield_expression) => {
+                self.infer_yield_expression(yield_expression);
+                Type::Unknown
+            }
+            ast::Expr::YieldFrom(yield_from) => {
+                self.infer_yield_from_expression(yield_from);
+                Type::Unknown
+            }
+            ast::Expr::Compare(compare) => {
+                self.infer_compare_expression(compare);
+                Type::Unknown
+            }
+            ast::Expr::Call(call_expr) => {
+                self.infer_call_expression(call_expr);
+                Type::Unknown
+            }
+            ast::Expr::FString(fstring) => {
+                self.infer_fstring_expression(fstring);
+                Type::Unknown
+            }
+            ast::Expr::Attribute(attribute) => {
+                self.infer_attribute_expression(attribute);
+                Type::Unknown
+            }
+            // TODO: this may be a place we need to revisit with special forms.
+            ast::Expr::Subscript(subscript) => {
+                self.infer_subscript_expression(subscript);
+                Type::Unknown
+            }
+            ast::Expr::Starred(starred) => {
+                self.infer_starred_expression(starred);
+                Type::Unknown
+            }
+            ast::Expr::List(list) => {
+                self.infer_list_expression(list);
+                Type::Unknown
+            }
+            ast::Expr::Tuple(tuple) => {
+                self.infer_tuple_expression(tuple);
+                Type::Unknown
+            }
+            ast::Expr::Slice(slice) => {
+                self.infer_slice_expression(slice);
+                Type::Unknown
+            }
+
+            // Literals do not have meaningful values in the annotation expression context. However,
+            // we will we want to handle these differently when working with special forms, since
+            // `"hello"` is not valid in an annotation expression but `Literal["hello"]` is.
```

Comment:
This isn't _quite_ right -- quoted annotations can be used for forward references in Python (something I'm currently struggling with in https://github.com/astral-sh/ruff/pull/12951), so `"hello"` might actually be a valid type annotation if `hello` resolves to a symbol in scope. See https://peps.python.org/pep-0484/#forward-references. (But you're correct that all of the others are not generally valid in type annotations. Except for inside `Literal` slices. And also `Ellipsis`, that also has some special meaning in type annotations 😄.)

Reply:
Oh, right, I forgot about that part (even though I was mucking around with it just yesterday)! Will update. I was wondering about `Ellipsis`, too. I will note that it is a “TODO” rather than a “nothing to see here” and go look up what it does.

Reply:
Did that in a3d82a8f.